### PR TITLE
Add kqueue-based uvar notifier for BSD

### PR DESCRIPTION
**Pull request by mqudsi**
Tested under FreeBSD 13.3.

This should also (theoretically) work under all versions of macOS, and has some benefits over the current notifyd choice. (No clue if it actually works there yet; CI might shed light on that!)

Mutex is used because of the non-mut `notification_fd_became_readable()` `&self` reference, but contention is not expected.

### CODE REVIEW

**Comment by mqudsi** [comment]
Looks like the kqueue notifier works on macOS as the corresponding rust test should be enabled on that platform (though not used).

FYI as a note for other BSDs in case this doesn't work out-of-the-box: from my reading, `NOTE_EXTEND` is *sometimes* `NOTE_TRUNC` on non-FreeBSD BSDs, but my gut instinct is that this applies when monitoring file fds, not dir fds.

**Comment by faho** [comment]
My question here is the same I have about all new notifiers: What is the advantage of this one? Does it mean we can drop any of the others?

**Comment by mqudsi** [comment]
> What are those benefits? 

I'm far from a macOS guru, but from my previous work with notifyd and kqueue more than a decade ago and as I recall the situation, notifyd is built on top of kqueue and extends it with a much more usable api. kqueue is a great api and it's not until io_uring that Linux had a proper alternative to it, but its fs monitoring operation was designed with watching single files rather than directory trees in mind; when an event happens in a directory, you are only told that an event happened but not given any further information than that (i.e. which file, what changed, etc). notifyd is a userland daemon that registers basic watches with the kernel (via kqueue itself?) and receives those basic change notifications, then provides downstream callers with more fleshed-out info about the changes that took place, at least as I understand and remember it.

But since notifications are basically enqueued for processing by notifyd, there is latency until notifyd handles the notifications and then because the notifyd api in a downstream app like fish effectively communicate via IPC with the notifyd daemon to receive those changes, there's additional overhead and latency when handling the results. (For example, we have to account for that extra latency [in our tests](https://github.com/fish-shell/fish-shell/blob/c9bc04f2745a3df18937844d9f13ada8aa59ca45/src/universal_notifier/test_helpers.rs#L52-L54).)

None of this is a big deal, but since we're already rebuilding the file state detection mechanism ourselves for correct BSD support (the inode and metadata comparisons), I only suggest that we might coalesce on kqueue uvar notifications for all platforms that support it (BSD + macOS) but it's just a suggestion and food for thought.

As for whether to keep around the notifyd code in general, that's above my pay grade as I don't use macOS on the daily and others may have better insight into the code and its maintenance.

An obvious "improvement" in the short term would be to just make the following change to the code in this branch, but whether it would ever see the new branch taken or not I don't know; I know that if the notifyd daemon is stopped or broken for whatever reason, the notifyd api breaks, but I don't know if that's something that actually happens (whereas I can tell you that if you were to rely on the search indexer on Windows to retrieve indexed metadata, *that* is constantly going down for one reason or the other):

```diff
﻿diff --git a/src/universal_notifier/mod.rs b/src/universal_notifier/mod.rs
index b0626b630..3c67fcb8c 100644
--- a/src/universal_notifier/mod.rs
+++ b/src/universal_notifier/mod.rs
@@ -56,17 +56,17 @@ pub fn create_notifier() -> Box<dyn UniversalNotifier> {
     #[cfg(target_os = "macos")]
     if let Some(notifier) = notifyd::NotifydNotifier::new() {
         return Box::new(notifier);
     }
     #[cfg(any(target_os = "android", target_os = "linux"))]
     if let Some(notifier) = inotify::InotifyNotifier::new() {
         return Box::new(notifier);
     }
-    #[cfg(bsd)]
+    #[cfg(any(bsd, target_os = "macos"))]
     if let Some(notifier) = kqueue::KqueueNotifier::new() {
         return Box::new(notifier);
     }
     Box::new(NullNotifier)
 }
```

(i.e. fall back to kqueue if notifyd fails for whatever reason)

**Review by ridiculousfish — APPROVED** [review]
Approach looks good - nice work. Some things to fix before merging, also there's dead code warnings on macOS.

Regarding using this implementation for macOS - let's not, at least initially. There's definitely tradeoffs with both implementations, but the `notifyd` approach is simple, supported since forever, and what we've been using for years. (Note it's based on Mach messages, not kqueue). We can re-evaluate in the future.

File: src/universal_notifier/kqueue.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        Some(self.kq.as_fd().as_raw_fd())
+    }
+
+    /// The notification_fd is readable; drain it.
+    /// Returns true if a notification is considered to have been posted.
+    fn notification_fd_became_readable(&self, fd: RawFd) -> bool {
+        let mut events = [std::mem::MaybeUninit::<KEvent>::uninit(); 1];
+        // Use a zero timeout just in case poll/select returned incorrectly/spuriously, but really
+        // that should never happen.
+        let timeout = libc::timespec {
```
[the remark this answers is not in the document]
Comment (faho) [reply src/universal_notifier/kqueue.rs:122]: I still believe that we're misunderstanding the time_t thing entirely. As I understand it, there is nothing for us to do.

**Pull request merged**